

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Quick start &mdash; stereopy v0.1 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "document", "processHtmlClass": "math|output_area"}}</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Examples" href="Examples.html" />
    <link rel="prev" title="stereo.plots.plot_heatmap_marker_genes" href="../api/stereo.plots.plot_heatmap_marker_genes.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> stereopy
          

          
          </a>

          
            
            
              <div class="version">
                v0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../General/Installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Read-the-raw-data-of-Stereo-Seq-(bin1)">Read the raw data of Stereo-Seq (bin1)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Preprocessing">Preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Quality-control">Quality control</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Filtering">Filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Normalization">Normalization</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Dimensionality-reduction">Dimensionality reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Clustering">Clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Find-marker-genes">Find marker genes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Hypothesis-test">Hypothesis test</a></li>
<li class="toctree-l3"><a class="reference internal" href="#spatial-lag-model">spatial lag model</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#spatial-pattern-score">spatial pattern score</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Examples.html">Examples</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">stereopy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Quick start</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/Tutorials/quick_start.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Quick-start">
<h1>Quick start<a class="headerlink" href="#Quick-start" title="Permalink to this headline">¶</a></h1>
<p>For example, we use the spatial expression matrix of Bin1 of Stereo-Seq in the mouse olfactory bulb for the use of <strong>stereopy</strong> tools for downstream analysis.</p>
<p>The format of the original expression matrix of Bin1 is as follows: x, y are the spatial position of the gene in the tissue section, and count is the number of gene expression.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 35%" />
<col style="width: 18%" />
<col style="width: 18%" />
<col style="width: 29%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>GeneID</p></th>
<th class="head"><p>x</p></th>
<th class="head"><p>y</p></th>
<th class="head"><p>count</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Gene1</p></td>
<td><p>121</p></td>
<td><p>200</p></td>
<td><p>2</p></td>
</tr>
<tr class="row-odd"><td><p>Gene2</p></td>
<td><p>234</p></td>
<td><p>300</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-even"><td><p>…</p></td>
<td><p>…</p></td>
<td><p>…</p></td>
<td><p>…</p></td>
</tr>
<tr class="row-odd"><td><p>Gene n</p></td>
<td><p>234</p></td>
<td><p>300</p></td>
<td><p>1</p></td>
</tr>
</tbody>
</table>
<div class="section" id="Read-the-raw-data-of-Stereo-Seq-(bin1)">
<h2>Read the raw data of Stereo-Seq (bin1)<a class="headerlink" href="#Read-the-raw-data-of-Stereo-Seq-(bin1)" title="Permalink to this headline">¶</a></h2>
<p>The resolution of Stereo-Seq sequencing technology is up to nanometer level. The nanopore radius of Bin1 is 250 nm.</p>
<p>Generally, the <strong>bin_size</strong> parameter is set to combine the nanopores within a certain range into a bin unit for analysis. tthe side length of the bin unit is approximately 14 microns when <strong>bin_size</strong> is set to 20.</p>
<p>After merging according to a certain <strong>bin_size</strong>, we will uniformly generate the data into the format of AnnData for downstream analysis. For details of AnnData, see <em>https://scanpy.readthedocs.io/en/latest/usage-principles.html#anndata</em></p>
<p>In addition, stereopy can support reading 10x data, as detailed in the API documentation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>import warnings

warnings.filterwarnings(&#39;ignore&#39;)
import stereo as st

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/data/workspace/install/Anaconda3/envs/st/lib/python3.9/site-packages/geopandas/_compat.py:106: UserWarning: The Shapely GEOS version (3.8.0-CAPI-1.13.1 ) is incompatible with the GEOS version PyGEOS was compiled with (3.9.0-CAPI-1.16.2). Conversions between both will be slow.
  warnings.warn(
/data/workspace/install/Anaconda3/envs/st/lib/python3.9/site-packages/libpysal/examples/remotes.py:26: UserWarning: Remote data sets not available. Check connection.
  warnings.warn(&#34;Remote data sets not available. Check connection.&#34;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>mouse_data_path = &#39;/data/workspace/st/data/stereopy_test/mouse/DP8400013846TR_F5.gem&#39;
andata = st.io.read_stereo_data(mouse_data_path, bin_size=100)
print(andata)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2021-05-27 11:10:21 Spateo INFO: the martrix has 3347 bins, and 28553 genes.
2021-05-27 11:10:26 Spateo INFO: the size of matrix is 4.470348358154297e-08 G.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 3347 × 28553
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
</div>
<div class="section" id="Preprocessing">
<h2>Preprocessing<a class="headerlink" href="#Preprocessing" title="Permalink to this headline">¶</a></h2>
<p>Data preprocessing includes three modules: <strong>Quality control</strong>, <strong>Filtering</strong> and <strong>Normalization</strong>.</p>
<div class="section" id="Quality-control">
<h3>Quality control<a class="headerlink" href="#Quality-control" title="Permalink to this headline">¶</a></h3>
<p>The quality control module mainly calculates the data quality distribution of the original data, and mainly counts three indicators: - the number of genes expressed in the count matrix. (n_genes_by_counts) - the total counts per cell. (total_counts) - the percentage of counts in mitochondrial genes. (pct_counts_mt)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>andata = st.preprocess.cal_qc(andata=andata)
</pre></div>
</div>
</div>
<p>Show the violin figure of the qc distribution.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plt.plot_violin_distribution(andata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_8_0.png" src="../_images/Tutorials_quick_start_8_0.png" />
</div>
</div>
<p>Show the spatial scatter figure of the qc distribution.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plt.plot_spatial_distribution(andata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_10_0.png" src="../_images/Tutorials_quick_start_10_0.png" />
</div>
</div>
</div>
<div class="section" id="Filtering">
<h3>Filtering<a class="headerlink" href="#Filtering" title="Permalink to this headline">¶</a></h3>
<p>Generally, we will filter the <code class="docutils literal notranslate"><span class="pre">bin</span> <span class="pre">unit</span></code> based on the three quality control indicators total_count, n_genes_by_counts and pct_counts_MT.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.preprocess.filter_cells(adata=andata, min_gene=200, n_genes_by_counts=3, pct_counts_mt=5, inplace=True)
</pre></div>
</div>
</div>
<p>In addition, we also provide <code class="docutils literal notranslate"><span class="pre">st.preprocess.filter_genes</span></code> function to filter genes and <code class="docutils literal notranslate"><span class="pre">st.preprocess.filter_coordinates</span></code> function to filer bin unit by its coordinate.</p>
<p>For detailed instructions, please refer to its api document.</p>
</div>
<div class="section" id="Normalization">
<h3>Normalization<a class="headerlink" href="#Normalization" title="Permalink to this headline">¶</a></h3>
<p>In this module, we mainly provide three common standardization methods.</p>
<ul class="simple">
<li><p>normalize_total</p></li>
<li><p>quantile</p></li>
<li><p>zscore</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>normalizer = st.preprocess.Normalizer(data=andata, method=&#39;normalize_total&#39;, inplace=True, target_sum=10000)
normalizer.fit()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 1.9984013 , 0.        , ..., 0.99920064, 0.        ,
        0.        ],
       ...,
       [0.        , 0.86737794, 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 1.0563009 , 0.        ,
        0.        ]], dtype=float32)
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Dimensionality-reduction">
<h2>Dimensionality reduction<a class="headerlink" href="#Dimensionality-reduction" title="Permalink to this headline">¶</a></h2>
<p>By default, PCA is used to reduce dimensionality.</p>
<p>In addition, it supports <code class="docutils literal notranslate"><span class="pre">tsen</span></code>, <code class="docutils literal notranslate"><span class="pre">umap</span></code>, <code class="docutils literal notranslate"><span class="pre">factor</span> <span class="pre">analysis</span></code>, <code class="docutils literal notranslate"><span class="pre">low</span> <span class="pre">variance</span></code> and other dimensionality reduction methods.See the API related documentation for details.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>dim_reduce = st.tools.DimReduce(andata=andata, method=&#39;pca&#39;, n_pcs=30, min_variance=0.01, n_iter=250, n_neighbors=10, min_dist=0.3, inplace=False, name=&#39;dim_reduce&#39;)
dim_reduce.fit()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>import pickle

pickle.dump(andata, open(&#39;./output/quick_start.pk&#39;, &#39;wb&#39;))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>import warnings
warnings.filterwarnings(&#39;ignore&#39;)
import pickle
import stereo as st


andata = pickle.load(open(&#39;./output/quick_start.pk&#39;, &#39;rb&#39;))
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plt.plot_dim_reduce(andata, plot_key=[&#39;Cst6&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_21_0.png" src="../_images/Tutorials_quick_start_21_0.png" />
</div>
</div>
</div>
<div class="section" id="Clustering">
<h2>Clustering<a class="headerlink" href="#Clustering" title="Permalink to this headline">¶</a></h2>
<p>Currently we provide two common clustering methods, including leiden and louvain.</p>
<p>In this tool, you can re-run the normalization method before clustering if the <code class="docutils literal notranslate"><span class="pre">normalize_method</span></code> parameter is not None. Then by default, we use pca to reduce the dimensionality of the new normalization method, and then using the top 30 pcs to run clustering.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>cluster = st.tools.Clustering(data=andata, method=&#39;leiden&#39;, dim_reduce_key=&#39;dim_reduce&#39;, n_neighbors=30, name=&#39;clustering&#39;)
cluster.fit()
</pre></div>
</div>
</div>
<p>Show the spatial distribution of the clusering result.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plots.plot_spatial_cluster(adata=andata, cluster_names=[&#39;clustering&#39;], pos_key=&#39;spatial&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_25_0.png" src="../_images/Tutorials_quick_start_25_0.png" />
</div>
</div>
<p>It can also show the dimensionality reduction spatial distribution of the clustering result if the <code class="docutils literal notranslate"><span class="pre">pos_key</span></code> is set to be dimensionality reduction tool’s name.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plots.plot_spatial_cluster(adata=andata, cluster_names=[&#39;clustering&#39;], pos_key=&#39;dim_reduce&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_27_0.png" src="../_images/Tutorials_quick_start_27_0.png" />
</div>
</div>
</div>
<div class="section" id="Find-marker-genes">
<h2>Find marker genes<a class="headerlink" href="#Find-marker-genes" title="Permalink to this headline">¶</a></h2>
<p>By default, the <code class="docutils literal notranslate"><span class="pre">.X</span></code> attribute of AnnData is used to compute the result. We support two tools to find marker genes, including <code class="docutils literal notranslate"><span class="pre">st.tools.FindMarker</span></code> and <code class="docutils literal notranslate"><span class="pre">st.tools.SpatialLag</span></code>.</p>
<div class="section" id="Hypothesis-test">
<h3>Hypothesis test<a class="headerlink" href="#Hypothesis-test" title="Permalink to this headline">¶</a></h3>
<p>In <code class="docutils literal notranslate"><span class="pre">st.tools.FindMarker</span></code>, it provides a hypothesis test to analyze differences between two groups. It support two methods including <code class="docutils literal notranslate"><span class="pre">t-test</span></code> and <code class="docutils literal notranslate"><span class="pre">wicoxon</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>marker = st.tools.FindMarker(data=andata, cluster=&#39;clustering&#39;, corr_method=&#39;bonferroni&#39;, method=&#39;t-test&#39;, name=&#39;marker_test&#39;)
marker.fit()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Find marker gene: 100%|██████████| 10/10 [00:27&lt;00:00,  2.75s/it]
2021-05-28 11:10:57 Spateo WARNING: your AnnData.uns[&#39;marker_test&#39;] is used by others, replacing it with a StereoResult.
</pre></div></div>
</div>
<p>Then, show the score of top 10 marker genes for the each group.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plt.plot_marker_genes(adata=andata, n_genes=10, find_maker_name=&#39;marker_test&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_33_0.png" src="../_images/Tutorials_quick_start_33_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>andata.uns[&#39;marker_test&#39;][&#39;1.vs.rest&#39;].top_k_marker(top_k_genes=10, ascend=False, sort_key=&#39;scores&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>genes</th>
      <th>scores</th>
      <th>pvalues</th>
      <th>pvalues_adj</th>
      <th>log2fc</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>9760</th>
      <td>Gm10800</td>
      <td>21.101646</td>
      <td>2.100798e-74</td>
      <td>5.998408e-70</td>
      <td>13.625463</td>
    </tr>
    <tr>
      <th>26395</th>
      <td>Gm20388</td>
      <td>13.959654</td>
      <td>2.427094e-38</td>
      <td>6.930082e-34</td>
      <td>17.820795</td>
    </tr>
    <tr>
      <th>23235</th>
      <td>Gm10801</td>
      <td>13.882109</td>
      <td>6.907171e-38</td>
      <td>1.972204e-33</td>
      <td>4.967365</td>
    </tr>
    <tr>
      <th>19542</th>
      <td>Auts2</td>
      <td>13.856010</td>
      <td>1.526443e-37</td>
      <td>4.358452e-33</td>
      <td>17.481253</td>
    </tr>
    <tr>
      <th>13176</th>
      <td>Gm48099</td>
      <td>13.152459</td>
      <td>8.335549e-35</td>
      <td>2.380049e-30</td>
      <td>7.280899</td>
    </tr>
    <tr>
      <th>11121</th>
      <td>Jarid2</td>
      <td>12.927252</td>
      <td>8.365243e-34</td>
      <td>2.388528e-29</td>
      <td>12.399784</td>
    </tr>
    <tr>
      <th>25125</th>
      <td>Gm26870</td>
      <td>12.505767</td>
      <td>8.863599e-32</td>
      <td>2.530823e-27</td>
      <td>3.556817</td>
    </tr>
    <tr>
      <th>13377</th>
      <td>Dlg2</td>
      <td>11.638679</td>
      <td>4.886943e-28</td>
      <td>1.395369e-23</td>
      <td>15.900744</td>
    </tr>
    <tr>
      <th>23668</th>
      <td>Parp4</td>
      <td>11.477016</td>
      <td>1.525292e-27</td>
      <td>4.355165e-23</td>
      <td>3.282708</td>
    </tr>
    <tr>
      <th>20491</th>
      <td>Gm21738</td>
      <td>11.456015</td>
      <td>2.169239e-27</td>
      <td>6.193828e-23</td>
      <td>2.914225</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Show the heatmap plot of top 10 marker genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plt.plot_heatmap_marker_genes(adata=andata, cluster_name=&#39;clustering&#39;, fine_maker_name=&#39;marker_test&#39;, num_show_gene=8)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_36_0.png" src="../_images/Tutorials_quick_start_36_0.png" />
</div>
</div>
</div>
<div class="section" id="spatial-lag-model">
<h3>spatial lag model<a class="headerlink" href="#spatial-lag-model" title="Permalink to this headline">¶</a></h3>
<p>By default, the tool calculate all genes’s lag coefficient, which takes some time.</p>
<p>If you only want to see the genes you are interested in, you can set the genes through the parameter <code class="docutils literal notranslate"><span class="pre">gene</span></code> .</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>lag = st.tools.SpatialLag(data=andata, cluster=&#39;clustering&#39;)
lag.fit()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
performing GM_lag_model and assign coefficient and p-val to cell type: 28553it [31:55, 14.91it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SpatialLagResult of spatial_lag.
  params: {&#39;data&#39;: AnnData object with n_obs × n_vars = 3273 × 28553
    obs: &#39;total_counts&#39;, &#39;pct_counts_mt&#39;, &#39;n_genes_by_counts&#39;, &#39;clustering&#39;
    uns: &#39;dim_reduce&#39;, &#39;clustering&#39;, &#39;marker_test&#39;, &#39;spatial_lag&#39;
    obsm: &#39;spatial&#39;, &#39;method&#39;: &#39;gm_lag&#39;, &#39;name&#39;: &#39;spatial_lag&#39;, &#39;cluster&#39;: &#39;clustering&#39;, &#39;genes&#39;: None, &#39;random_drop&#39;: True, &#39;drop_dummy&#39;: None, &#39;n_neighbors&#39;: 8}
    result: a DataFrame which has `genes`,`lag_coeff`, `lag_zstat`, `lag_pval` columns.
    the shape is: (28553, 36)
</pre></div></div>
</div>
<p>Show the result of spatial lag model.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>andata.uns[&#39;spatial_lag&#39;].score
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>const_lag_coeff</th>
      <th>const_lag_zstat</th>
      <th>const_lag_pval</th>
      <th>1_lag_coeff</th>
      <th>1_lag_zstat</th>
      <th>1_lag_pval</th>
      <th>6_lag_coeff</th>
      <th>6_lag_zstat</th>
      <th>6_lag_pval</th>
      <th>0_lag_coeff</th>
      <th>...</th>
      <th>4_lag_pval</th>
      <th>9_lag_coeff</th>
      <th>9_lag_zstat</th>
      <th>9_lag_pval</th>
      <th>7_lag_coeff</th>
      <th>7_lag_zstat</th>
      <th>7_lag_pval</th>
      <th>W_log_exp_lag_coeff</th>
      <th>W_log_exp_lag_zstat</th>
      <th>W_log_exp_lag_pval</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Iyd</th>
      <td>0.0</td>
      <td>13.46056</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.056549</td>
      <td>0.954905</td>
      <td>-0.0</td>
      <td>-3.907563</td>
      <td>0.000093</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.000542</td>
      <td>-0.0</td>
      <td>-11.054426</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-11.960677</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>126180040290664272.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Rarb</th>
      <td>0.0</td>
      <td>16.521538</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-11.816965</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-3.050368</td>
      <td>0.002286</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.107645</td>
      <td>-0.0</td>
      <td>-4.021895</td>
      <td>0.000058</td>
      <td>-0.0</td>
      <td>-3.242075</td>
      <td>0.001187</td>
      <td>1.0</td>
      <td>6399440673649731.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Irx2</th>
      <td>-0.0</td>
      <td>-2.27271</td>
      <td>0.023044</td>
      <td>0.0</td>
      <td>7.487105</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-5.804079</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.682396</td>
      <td>0.0</td>
      <td>0.930905</td>
      <td>0.351903</td>
      <td>-0.0</td>
      <td>-1.573045</td>
      <td>0.115708</td>
      <td>1.0</td>
      <td>114601771562122768.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3300002A11Rik</th>
      <td>0.0</td>
      <td>18.742546</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-12.819551</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-16.572084</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-9.882764</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-8.218894</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>106285634107647536.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Gm12946</th>
      <td>0.0</td>
      <td>2.276698</td>
      <td>0.022804</td>
      <td>0.0</td>
      <td>7.942856</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-8.335728</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.002284</td>
      <td>-0.0</td>
      <td>-16.062968</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-13.694554</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>201367820094472832.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Cst6</th>
      <td>0.0</td>
      <td>10.516882</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.243903</td>
      <td>0.807306</td>
      <td>0.0</td>
      <td>0.393624</td>
      <td>0.693858</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.000001</td>
      <td>-0.0</td>
      <td>-4.51893</td>
      <td>0.000006</td>
      <td>0.0</td>
      <td>3.619165</td>
      <td>0.000296</td>
      <td>1.0</td>
      <td>50826626299372016.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Ascl3</th>
      <td>0.0</td>
      <td>9.524975</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-3.055332</td>
      <td>0.002248</td>
      <td>0.0</td>
      <td>12.598778</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.000289</td>
      <td>-0.0</td>
      <td>-5.230889</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-4.320097</td>
      <td>0.000016</td>
      <td>1.0</td>
      <td>100451245031858672.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Gmpr</th>
      <td>0.0</td>
      <td>9.391973</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-3.804836</td>
      <td>0.000142</td>
      <td>0.0</td>
      <td>3.066049</td>
      <td>0.002169</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-1.37376</td>
      <td>0.169516</td>
      <td>0.0</td>
      <td>14.603169</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>37505532397988568.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Gm28863</th>
      <td>0.0</td>
      <td>5.008487</td>
      <td>0.000001</td>
      <td>0.0</td>
      <td>10.167506</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-1.164742</td>
      <td>0.244123</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.151654</td>
      <td>0.0</td>
      <td>2.059288</td>
      <td>0.039467</td>
      <td>-0.0</td>
      <td>-3.238117</td>
      <td>0.001203</td>
      <td>1.0</td>
      <td>78119007377666144.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Gm39091</th>
      <td>0.0</td>
      <td>18.136187</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-16.538562</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-32.052758</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-12.747961</td>
      <td>0.0</td>
      <td>-0.0</td>
      <td>-12.722034</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>489270358257124672.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>28553 rows × 36 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>st.plt.plot_heatmap_marker_genes(adata=andata, cluster_name=&#39;clustering&#39;, fine_maker_name=&#39;spatial_lag&#39;, num_show_gene=8)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;1&#39;, &#39;6&#39;, &#39;0&#39;, &#39;3&#39;, &#39;2&#39;, &#39;5&#39;, &#39;8&#39;, &#39;4&#39;, &#39;9&#39;, &#39;7&#39;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Tutorials_quick_start_42_1.png" src="../_images/Tutorials_quick_start_42_1.png" />
</div>
</div>
</div>
</div>
<div class="section" id="spatial-pattern-score">
<h2>spatial pattern score<a class="headerlink" href="#spatial-pattern-score" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>sps = st.tools.SpatialPatternScore(data=andata)
sps.fit()
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="Examples.html" class="btn btn-neutral float-right" title="Examples" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../api/stereo.plots.plot_heatmap_marker_genes.html" class="btn btn-neutral float-left" title="stereo.plots.plot_heatmap_marker_genes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Ping Qiu.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>